### PR TITLE
Update list-patterns.md

### PR DESCRIPTION
**Pull request by alrz**
Relates to the PR https://github.com/dotnet/roslyn/pull/57457

### CODE REVIEW

File: proposals/list-patterns.md [hunk 1]
```diff
@@ -69,7 +69,12 @@ As a result, an error is produced for something like `case [.., p]: case [p]:` b
 
 If a slice subpattern matches a list or a length value, subpatterns are treated as if they were a direct subpattern of the containing list. For instance, `[..[1, 2, 3]]` subsumes a pattern of the form `[1, 2, 3]`.
 
-`Length` or `Count` properties are assumed to always return a non-negative value, if and only if the type is *indexable*. For instance, the pattern `{ Length: -1 }` can never match an array. The behavior of a pattern-matching operation is undefined if this assumption doesn't hold.
+The following assumptions are made on the members being used:
+
+- `Length` or `Count` properties are assumed to always return a non-negative value, if and only if the type is *indexable*. For instance, the pattern `{ Length: -1 }` can never match an array.
+- The `Slice` method or the range indexer is assumed to be well-behaved, that is, the return value is never null and that it is a proper subslice of the containing list.
```
Comment (alrz) [inline proposals/list-patterns.md:75]: Note that this only affects the codegen and nullability exhaustiveness. 
We never produce an error if an unannotated type is matched with `null`.

**Comment by alrz** [comment]
@jcouv @333fred

File: proposals/list-patterns.md [hunk 2]
```diff
@@ -69,7 +69,12 @@ As a result, an error is produced for something like `case [.., p]: case [p]:` b
 
 If a slice subpattern matches a list or a length value, subpatterns are treated as if they were a direct subpattern of the containing list. For instance, `[..[1, 2, 3]]` subsumes a pattern of the form `[1, 2, 3]`.
 
-`Length` or `Count` properties are assumed to always return a non-negative value, if and only if the type is *indexable*. For instance, the pattern `{ Length: -1 }` can never match an array. The behavior of a pattern-matching operation is undefined if this assumption doesn't hold.
+The following assumptions are made on the members being used:
+
+- `Length` or `Count` properties are assumed to always return a non-negative value, if and only if the type is *indexable*. For instance, the pattern `{ Length: -1 }` can never match an array.
```
Comment (jcouv) [inline proposals/list-patterns.md:74]: Consider "The property that makes the type countable is assumed to ...".
In the scenario where both `Length` and `Count` are present, only `Length` gets the benefit of the assumption.

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 1)

**Comment by jcouv** [comment]
Thanks for the follow-up (remembering to document the assumptions) :-)

**Comment by alrz** [comment]
>(remembering to document the assumptions)

Almost.

**Comment by jcouv** [comment]
Oops. I may have merged too soon.
Did we actually assume that Slice is not-null? That's not what I recall.

**Comment by jcouv** [comment]
Notes say: "If a Slice method is annotated, we will treat it as potentially returning null for the purposes of subsumption and exhaustiveness."
https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-10-20.md#open-questions-in-list-patterns

**Comment by alrz** [comment]
No, that's not what's been discussed in LDM. This is a turnaround. 

I just realized this directly affects subsumption checking for nested slices so I figured we need to revisit.

**Comment by alrz** [comment]
Details to discuss: https://github.com/dotnet/csharplang/pull/5364#discussion_r739507602

**Pull request merged**